### ISSUE TITLE
SAM_RS：Boundary Loss Correction

### ISSUE THREAD

**Issue by WanZhan-lucky** [sstary/SSRS#52]
Dear author, hello！In the SAM_RS folder in your code, find the BoundaryLoss and one_hot functions in utils.py, as follows：

```
def one_hot(label, n_classes, requires_grad=True):
    """Return One Hot Label"""
    one_hot_label = torch.eye(
        n_classes, device='cuda', requires_grad=requires_grad)[label]
    one_hot_label = one_hot_label.transpose(1, 3).transpose(2, 3)

    return one_hot_label
```

```
class BoundaryLoss(nn.Module):
    def __init__(self, theta0=3, theta=5):
        super().__init__()

        self.theta0 = theta0
        self.theta = theta

    def forward(self, pred, gt):
        """
        Input:
            - pred: the output from model (before softmax)
                    shape (N, C, H, W)
            - gt: ground truth map
                    shape (N, H, w)
        Return:
            - boundary loss, averaged over mini-bathc
        """
        n, c, _, _ = pred.shape
        # softmax so that predicted map can be distributed in [0, 1]
        pred = torch.softmax(pred, dim=1)
        # one-hot vector of ground truth
        one_hot_gt = one_hot(gt, c)
        # boundary map
        gt_b = F.max_pool2d(
            1 - one_hot_gt, kernel_size=self.theta0, stride=1, padding=(self.theta0 - 1) // 2)
        gt_b -= 1 - one_hot_gt

        pred_b = F.max_pool2d(
            1 - pred, kernel_size=self.theta0, stride=1, padding=(self.theta0 - 1) // 2)
        pred_b -= 1 - pred

        # extended boundary map
        gt_b_ext = F.max_pool2d(
            gt_b, kernel_size=self.theta, stride=1, padding=(self.theta - 1) // 2)

        pred_b_ext = F.max_pool2d(
            pred_b, kernel_size=self.theta, stride=1, padding=(self.theta - 1) // 2)

        # reshape
        gt_b = gt_b.view(n, c, -1)
        pred_b = pred_b.view(n, c, -1)
        gt_b_ext = gt_b_ext.view(n, c, -1)
        pred_b_ext = pred_b_ext.view(n, c, -1)

        # Precision, Recall
        P = torch.sum(pred_b * gt_b_ext, dim=2) / (torch.sum(pred_b, dim=2) + 1e-7)
        R = torch.sum(pred_b_ext * gt_b, dim=2) / (torch.sum(gt_b, dim=2) + 1e-7)

        # Boundary F1 Score
        BF1 = 2 * P * R / (P + R + 1e-7)

        # summing BF1 Score for each class and average over mini-batch
        loss = torch.mean(1 - BF1)

        return loss

```
In addition, I found the generation and preprocessing of boundry in SAM_utils, as follows

```
 BoundaryPrior [np.where( BoundaryPrior >0) ]=1
    BoundaryPrior_index=np.where(BoundaryPrior >0)
    Objects_first_few[BoundaryPrior_index]= 0  
    BoundaryPrior_output [np.where( BoundaryPrior >0) ]=255
    BoundaryPrior_output = BoundaryPrior_output.astype(np.uint8) 
    return BoundaryPrior_output,Objects_first_few  
```

```
boundary = np.asarray(io.imread(self.boundary_files[random_idx])) / 255
            boundary = boundary.astype(np.int64)
```

      The png file of the boundary, all boundaries are 255, and the background is 0. During the training preprocessing, it is divided by 255, so all 1s are boundaries and 0s are backgrounds. Next, we use LoveA Urban as training, c=7, and then we look at the BoundayLoss function and the one_hot function. 
     gt is a binary image (1, H, W), where 1 represents the boundary and 0 represents the background. But in one_hot encoding, it converts gt into a one-hot encoding matrix (1, 7, H, W) with 7 channels, and each category occupies one channel independently. gt only contains 0 and 1, so after one-hot encoding, only the first two channels will have values ​​(channel 0 represents the background, channel 1 represents the boundary), and the other channels are all 0. Since the softmax output of pred is a probability distribution of 7 channels, multiplying it channel by channel with the one-hot encoded gt is not very reasonable in my opinion, because it only cares about the boundary and not the category.
     Therefore, I think merging pred channels: pred.sum(dim=1, keepdim=True) merges all categories of channels into one channel (i.e., boundary probability) so that the boundary loss can be calculated without caring about the specific category. .The specific code modification is to remove the One_hot function
pred_b = F.max_pool2d(1 - pred.sum(dim=1, keepdim=True), kernel_size=self.theta0, stride=1, padding=(self.theta0 - 1) // 2)
pred_b = pred_b - (1 - pred.sum(dim=1, keepdim=True)) 

     In other words, your code focuses on calculating the boundary loss of each category, but I don’t see the boundary information of different categories in Boundary. Finally, I highly appreciate your work and hope to get a reply as soon as possible if you have the right time. Thank you very much.

**Comment by sstary** [comment]
Yes, you are right. 

We applied the Boundary loss as published in published work 'Boundary Loss for Remote Sensing Imagery Semantic Segmentation' and the code from https://github.com/yiskw713/boundary_loss_for_remote_sensing. 

After checking, we found that our gt is inconsistent with the original gt. Our boundary is produced by SAM, and there is no semantic information, so there are only 0 and 1. While the original boundary is extracted by the semantic label, with the value of 0,1,... c. This leads us to calculate actually only class 1 that is the building class. Building is a relatively large category, which may be why imperfect code could also achieve good results.

Thank you for providing us with a good solution. We can't change gt, so we can only take your solution and change the output properties of pred. However, it is not reasonable to use pred=pred.sum(dim=1, keepdim=True), which would cause the entire pred to become all 1s since there is pred = torch.softmax(pred, dim=1) before.

We will update the code and thank you for your contributions on this github.

**Comment by sstary** [comment]
We have corrected the mistake in boundary loss and provided a simple boundary extraction algorithm to get the model's boundary predictions from the prediction map. It will work with the SGB to calculate the boundary loss.

We use the trained model, get the output of a image and extract its boundaries. The following figure shows that the simple boundary extraction algorithm can generate the expected predicted boundary, so it is also expected to be used in the calculation of new boundary loss as follows:

![image](https://github.com/user-attachments/assets/d82db1bd-29b4-4ab6-9a38-e933d40b8317)

Thank you again for your help and I really hope more feedback from you.

**Comment by sstary** [comment]
We update it again. We get the class map from the prediction map and follow the original processing of producing the boundary map.
